numpyro/handlers.py

```diff
@@ -437,3 +438,74 @@ def process_message(self, msg):
 
         if value is not None:
             msg['value'] = value
+
+
+class do(Messenger):
+    """
+    Given a stochastic function with some sample statements
+    and a dictionary of values at names,
+    set the return values of those sites equal to the values
+    as if they were hard-coded to those values
+    and introduce fresh sample sites with the same names
+    whose values do not propagate.
+    Composes freely with :func:`~numpyro.handlers.condition`
+    to represent counterfactual distributions over potential outcomes.
+    See Single World Intervention Graphs [1] for additional details and theory.
+
+    This is equivalent to replacing `z = numpyro.sample("z", ...)` with `z = 1.`
+    and introducing a fresh sample site numpyro.sample("z", ...) whose value is not used elsewhere.
+
+    References
+
+    [1] `Single World Intervention Graphs: A Primer`,
+        Thomas Richardson, James Robins
+
+    :param fn: a stochastic function (callable containing Pyro primitive calls)
+    :param data: a ``dict`` mapping sample site names to interventions
+
+    **Example:**
+
+    .. doctest::
+
+      >>> import jax.numpy as jnp
+      >>> import numpyro
+      >>> from numpyro.handlers import do, trace, seed
+      >>> import numpyro.distributions as dist
+      >>> def model(x):
+      ...     s = numpyro.sample("s", dist.LogNormal())
+      ...     z = numpyro.sample("z", dist.Normal(x, s))
+      ...     return z ** 2
+      >>> intervened_model = handlers.do(model, data={"z": 1.})
+      >>> with trace() as exec_trace:
+      ...     z_square = seed(intervened_model, 0)(x)
+      >>> assert exec_trace['z']['value'] != 1.
+      >>> assert not exec_trace['z']['is_observed']
+      >>> assert not exec_trace['z'].get('stop', None)
+      >>> assert z_square == 1
+    """
+    def __init__(self, fn=None, data=None):
```

Comment:
I think this is almost done. We just need to change `data` to `param_map` here and throughout docs and tests. In NumPyro, we used `param_map` for `condition`, `substitute`,... instead of `data`.

@neerajprad Do you think it is better to change all the names to `data`? The benefit is it will match Pyro api, but we have to raise Deprecation error if users use `param_map`. I don't have a strong opinion on this.

Reply:
IIRC `param_map` was specifically in the context of `substitute` since we can't have a global param store. I'm not sure if that distinction is relevant now, and otherwise I think it is a good idea to keep parity with Pyro wherever possible. We can change the argument for condition (and substitute) in a separate PR.

Reply:
Agreed! I'll try to collect some differences that I know in a separate PR.

Reply:
@fehiepsi do you still want me to change `data` to `param_map` ?

Reply:
Ah, sorry, I think you can keep it because we are going to change `param_map` to `data` for other handlers.